Mathlib/Data/Set/Subset.lean

```diff
@@ -0,0 +1,270 @@
+/-
+Copyright (c) 2024 Miguel Marco. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Miguel Marco
+-/
+import Mathlib.Data.Set.Function
+import Mathlib.Data.Set.Functor
+import Mathlib.Lean.Expr.ExtraRecognizers
+
+/-!
+# Subsets as subtypes
+
+This file defines notation for sets in a type pulled down to sets in a subtype, and sets in
+a subtype, lifted up to sets in the ambient type.
+
+It also provides some related lemmas for convenience.
+
+## Notation
+
+Let `α` be a `Type`, `A B : Set α` two sets in `α`, and `C : Set ↑A` a set in the subtype `↑A`.
+
+- `A ↓∩ B` denotes `(Subtype.val ⁻¹' B : Set A)` (that is, `{x : ↑A | ↑x ∈ B}`).
+- `↑C` denotes `Subtype.val '' C` (that is, `{x : α | ∃ y ∈ C, ↑y = x}`).
+
+To access this notation, type `open Subset`.
+
+## Tags
+
+subsets
+-/
+
+open Set
+
+universe u
+
+variable {α β : Type u} {A B C: Set α} {D E : Set ↑A}
+variable {S : Set (Set α)} {T : Set (Set ↑A)} {i : β → Set α} {j : β → Set ↑A}
+
+namespace Subset
+
+/--
+Given two sets `A` and `B`, `A ↓∩ B` denotes `((↑) ⁻¹' B : Set A)`.
+-/
+scoped notation A:67 " ↓∩ " B:67  => ((↑) ⁻¹' B : Set A)
```

Comment:
Just to check, do we want the following to work?
```lean
variable (X : Set ℤ) (Y : Set ℕ)
#check Y ↓∩ X -- gives a Set Y
```
My first reaction is that we should try to lock the notation down more to be just for subtypes, but I'm not completely sure.

I've found that the following works to lock it down, and it also generates a delaborator for you.
```lean
scoped notation3 A:67 " ↓∩ " B:67 => (Subtype.val ⁻¹' B : Set (A : Set _))
```

Reply:
Good catch! No, I don't think we should allow `A` and `B` being of different types.

So... `notation3` works just as `notation` but produces a delaborator too?

Reply:
I see the docstring of `notation3` warns about its uncertain future. Would it be a good idea to keep it?

Reply:
`notation3` might have an uncertain future, but there will definitely be a future for the features it provides, and in any case there are no concrete plans to deprecate it. I hope to get `notation` to be able to generate a delaborator too one day.

Reply:
I see. I pushed a commit with your suggestion.

I also changed the names of the lemmas to `preimage_val_...` since that is the underlying type now. 

Still unsure on what the namespace (and file) should be called.